esmvaltool/diag_scripts/shared/_base.py

```diff
@@ -51,14 +53,85 @@ def get_diagnostic_filename(basename, cfg, extension='nc'):
     -------
     str:
         A valid path for saving a diagnostic data file.
-
     """
     return os.path.join(
         cfg['work_dir'],
         f"{basename}.{extension}",
     )
 
 
+def save_figure(basename, provenance, cfg, figure=None, close=True, **kwargs):
+    """Save a figure to file.
+
+    Parameters
+    ----------
+    basename: str
+        The basename of the file.
+    provenance: dict
+        The provenance record for the figure.
+    cfg: dict
+        Dictionary with diagnostic configuration.
+    figure: matplotlib.figure.Figure
+        Figure to save.
+    close: bool
+        Close the figure after saving.
+    **kwargs:
+        Keyword arguments to pass to :obj:`matplotlib.figure.Figure.savefig`.
+
+    See Also
+    --------
+    ProvenanceLogger: For an example provenance record that can be used
+        with this function.
+    """
+    if cfg.get('output_file_type') is None:
+        extensions = ('png', 'pdf')
+    elif isinstance(cfg['output_file_type'], str):
+        extensions = (cfg['output_file_type'], )
+    else:
+        extensions = cfg['output_file_type']
+
+    for ext in extensions:
+        filename = Path(cfg['plot_dir']) / ext / f"{basename}.{ext}"
+        filename.parent.mkdir(exist_ok=True)
+        logger.info("Plotting analysis results to %s", filename)
+        fig = plt if figure is None else figure
+        fig.savefig(filename, **kwargs)
+        with ProvenanceLogger(cfg) as provenance_logger:
+            provenance_logger.log(filename, provenance)
+
+    if close:
+        plt.close(figure)
+
+
+def save_data(basename, provenance, cfg, cube, **kwargs):
+    """Save the data used to create a plot to file.
+
+    Parameters
+    ----------
+    basename: str
+        The basename of the file.
+    provenance: dict
+        The provenance record for the data.
+    cfg: dict
+        Dictionary with diagnostic configuration.
+    cube: iris.cube.Cube
+        Data cube to save.
+    **kwargs:
+        Keyword arguments to pass to :obj:`iris.save`.
+
+    See Also
+    --------
+    ProvenanceLogger: For an example provenance record that can be used
+        with this function.
+    """
+    if cfg.get('write_netcdf', True):
+        filename = get_diagnostic_filename(basename, cfg)
+        logger.info("Saving analysis results to %s", filename)
+        iris.save(cube, target=filename, **kwargs)
```

Comment:
```suggestion
        kwargs['target'] = filename # to make sure filename is not used in kwargs
        iris.save(cube, **kwargs)
        logger.info("Wrote %s", filename)
```

Same here?

Reply:
Ignore the log, but would it make sense to add

```py
kwargs['target'] = filename
```
?

Reply:
Changed the function so it raises a ValueError if kwargs contains target, because this is not the intended use. Is that OK?

Reply:
Yes!